CIS Terminate (P/C) and CIS Create (P) procedures implemented

erbr-ot opened the pull request:
This PR implements control procedures for CIS Terminate for peripheral and central device and CIS Create for peripheral device.

Testing was done so far on Oticon specific version of Zephyr main in bsim (very rudimentary towards Packetcraft CIS Central using EDTT and Oticon App test framework) as well as by unit tests.

sjanc commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -543,39 +543,65 @@ static bool is_valid_disconnect_reason(uint8_t reason)
+#if defined(CONFIG_BT_CTLR_PERIPHERAL_ISO) || defined(CONFIG_BT_CTLR_CENTRAL_ISO)
+	if (IS_CIS_HANDLE(handle)) {
+#if !defined(CONFIG_BT_LL_SW_LLCP_LEGACY)
+		cis = ll_conn_iso_stream_get(handle);
```
can be null?

erbr-ot replied:
No, the get() will return a valid pointer in all cases, however perhaps there needs to be another check for validity of the cis.

sjanc replied:
OK, I confused this with ll_conn_iso_stream_get_by_group() and other helpers

erbr-ot replied:
Fixed this, needed to use a different API to get the cis. Thnx for noticing!

erbr-ot commented:
Fixed a minor issue in CIS Create unit test, and rebased

erbr-ot commented:
Rebased

erbr-ot commented:
rebase

erbr-ot commented:
rebased AND fixed issues

erbr-ot commented:
rebased

erbr-ot commented:
rebase

erbr-ot commented:
rebased

erbr-ot commented:
rebased

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_cc.c:
```diff
@@ -0,0 +1,603 @@
+	/* This is equivalent to:
+	 * exactly one bit set, and no bit set is rfu's
+	 */
+	return (phy == 1 || phy == 2 || phy == 4);
```
I'm sure we have some `#define`s for these PHY values somewhere :) Please use those

erbr-ot replied:
I specifically didn't use the PHY defines because I had seen it as an abstraction, but it is equivalent, so I'll just use the defines to avoid soar eyes ;-)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_cc.c:
```diff
@@ -0,0 +1,603 @@
+static uint8_t rp_cc_check_phy(struct ll_conn *conn, struct proc_ctx *ctx,
+					    struct pdu_data *pdu)
+{
+	if (!phy_valid(pdu->llctrl.cis_req.c_phy) || !phy_valid(pdu->llctrl.cis_req.p_phy)) {
```
```suggestion
	if (!phy_valid(pdu->llctrl.cis_req.c_phy) || 
	    !phy_valid(pdu->llctrl.cis_req.p_phy)) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_cc.c:
```diff
@@ -0,0 +1,603 @@
+		break;
+	}
+}
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_cc.c:
```diff
@@ -0,0 +1,603 @@
+		break;
+	}
+}
+
```
```suggestion
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_cc.c:
```diff
@@ -0,0 +1,603 @@
+	}
+}
+
+bool llcp_rp_cc_awaiting_reply(struct proc_ctx *ctx)
```
```suggestion
bool llcp_rp_cc_awaiting_reply(const struct proc_ctx *ctx)
```

erbr-ot replied:
Skipping this, as I will not do this change everywhere just now;-)

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -98,6 +104,36 @@ enum {
+ * @brief Stop and tear down a connected ISO stream
+ * This function may be called to tear down a CIS.
+ *
+ * @param cig_id         ID of specific ISO group
+ * @param cis_id	 ID of connected ISO stream to stop
+ * @param reason         Termination reason
```
```suggestion
 * @param cig_id         ID of specific ISO group
 * @param cis_id         ID of connected ISO stream to stop
 * @param reason         Termination reason
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -462,6 +512,13 @@ static void lp_comm_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
 }
 
+#if defined(CONFIG_BT_CTLR_CENTRAL_ISO) || defined(CONFIG_BT_CTLR_PERIPHERAL_ISO)
+static bool lp_cis_terminated(struct ll_conn *conn, struct proc_ctx *ctx)
```
```suggestion
static bool lp_cis_terminated(const struct ll_conn *conn)
```

`ctx` is unused

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -584,6 +652,17 @@ static void lp_comm_st_idle(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t
+			/* So we should start by requesting Terminate for the CIS in question */
+
+			/* Clear terminate ack flag, used to signal CIS Terminated */
+			conn->llcp.cis.terminate_ack = 0;
```
```suggestion
			conn->llcp.cis.terminate_ack = 0U;
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_common.c:
```diff
@@ -794,6 +881,15 @@ static void rp_comm_terminate(struct ll_conn *conn, struct proc_ctx *ctx)
 }
+
+#if defined(CONFIG_BT_CTLR_CENTRAL_ISO) || defined(CONFIG_BT_CTLR_PERIPHERAL_ISO)
+static void rp_comm_stop_cis(struct ll_conn *conn, struct proc_ctx *ctx)
```
```suggestion
static void rp_comm_stop_cis(const struct proc_ctx *ctx)
```
`conn` is unused

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_features.h:
```diff
@@ -132,22 +132,22 @@ static inline bool feature_cte_req(struct ll_conn *conn)
  */
 static inline bool feature_smi_rx(struct ll_conn *conn)
 {
-	return LL_FEAT_BIT_SMI_RX;
+	return LL_FEAT_BIT_SMI_RX != 0;
 }
```
`conn` unused?

erbr-ot replied:
yeap

erbr-ot replied:
No, will mark arg as unused, so as to keep header identical with other similars

Thalley replied:
I'll let this be up to you guys working in the controller, but at least I do not see the point of keeping the argument if it is unused, unless it's used with a specific callback function definition :)

erbr-ot replied:
Agree - will remove

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c:
```diff
@@ -828,3 +834,73 @@ void llcp_pdu_encode_cte_rsp(const struct proc_ctx *ctx, struct pdu_data *pdu)
 #endif /* CONFIG_BT_CTLR_DF_CONN_CTE_RSP */
+
+#if defined(CONFIG_BT_CTLR_PERIPHERAL_ISO)
+void llcp_pdu_decode_cis_req(struct proc_ctx *ctx, struct pdu_data *pdu)
```
```suggestion
void llcp_pdu_decode_cis_req(struct proc_ctx *ctx, const struct pdu_data *pdu)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_llcp_pdu.c:
```diff
@@ -828,3 +834,73 @@ void llcp_pdu_encode_cte_rsp(const struct proc_ctx *ctx, struct pdu_data *pdu)
+}
+#endif /* defined(CONFIG_BT_CTLR_PERIPHERAL_ISO) */
+#if defined(CONFIG_BT_CTLR_CENTRAL_ISO) || defined(CONFIG_BT_CTLR_PERIPHERAL_ISO)
+void llcp_pdu_encode_cis_terminate_ind(struct proc_ctx *ctx, struct pdu_data *pdu)
```
```suggestion
void llcp_pdu_encode_cis_terminate_ind(const struct proc_ctx *ctx, struct pdu_data *pdu)
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_peripheral_iso.c:
```diff
@@ -39,14 +45,19 @@
 #include "hal/debug.h"
 
-uint8_t ll_cis_accept(uint16_t handle)
+static struct ll_conn *ll_cis_reply_ok(uint16_t handle, uint8_t *error)
```
Function name implies a boolean return value, but that's not the case here. Suggest to rename

Thalley replied:
Why isn't it? `ll_cis_reply_ok` is added by this PR, isn't it?

erbr-ot replied:
Ah, now I see what you mean. Will rename

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -402,6 +407,79 @@ void helper_node_encode_cte_rsp(struct node_rx_pdu *rx, void *param)
+		   sizeof(struct pdu_data_llctrl_cis_req);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_CIS_REQ;
+
+	pdu->llctrl.cis_req.cig_id		=  p->cig_id;
+	pdu->llctrl.cis_req.cis_id		=  p->cis_id;
+	pdu->llctrl.cis_req.c_phy		=  p->c_phy;
+	pdu->llctrl.cis_req.p_phy		=  p->p_phy;
+	pdu->llctrl.cis_req.c_max_pdu		=  p->c_max_pdu;
+	pdu->llctrl.cis_req.p_max_pdu		=  p->p_max_pdu;
+	pdu->llctrl.cis_req.nse			=  p->nse;
+	pdu->llctrl.cis_req.p_bn		=  p->p_bn;
+	pdu->llctrl.cis_req.c_bn		=  p->c_bn;
+	pdu->llctrl.cis_req.c_ft		=  p->c_ft;
+	pdu->llctrl.cis_req.p_ft		=  p->p_ft;
+	pdu->llctrl.cis_req.iso_interval	=  p->iso_interval;
+	pdu->llctrl.cis_req.conn_event_count	= p->conn_event_count;
+	memcpy(pdu->llctrl.cis_req.c_max_sdu_packed, p->c_max_sdu_packed,
+	       sizeof(p->c_max_sdu_packed));
+	memcpy(pdu->llctrl.cis_req.p_max_sdu, p->p_max_sdu, sizeof(p->p_max_sdu));
+	memcpy(pdu->llctrl.cis_req.c_sdu_interval, p->c_sdu_interval, sizeof(p->c_sdu_interval));
+	memcpy(pdu->llctrl.cis_req.p_sdu_interval, p->p_sdu_interval, sizeof(p->p_sdu_interval));
+	memcpy(pdu->llctrl.cis_req.sub_interval, p->sub_interval, sizeof(p->sub_interval));
+	memcpy(pdu->llctrl.cis_req.cis_offset_min, p->cis_offset_min, sizeof(p->cis_offset_min));
+	memcpy(pdu->llctrl.cis_req.cis_offset_max, p->cis_offset_max, sizeof(p->cis_offset_max));
```
Please only use spaces to indent after a non-tab character.
Conventionally we (void) cast the return values of the stdlib functions (like `memcpy`) if the return value isn't used. 

Please also keep lines below 80 when possible. 

```suggestion
	pdu->llctrl.cis_req.cig_id              = p->cig_id;
	pdu->llctrl.cis_req.cis_id              = p->cis_id;
	pdu->llctrl.cis_req.c_phy               = p->c_phy;
	pdu->llctrl.cis_req.p_phy               = p->p_phy;
	pdu->llctrl.cis_req.c_max_pdu           = p->c_max_pdu;
	pdu->llctrl.cis_req.p_max_pdu           = p->p_max_pdu;
	pdu->llctrl.cis_req.nse                 = p->nse;
	pdu->llctrl.cis_req.p_bn                = p->p_bn;
	pdu->llctrl.cis_req.c_bn                = p->c_bn;
	pdu->llctrl.cis_req.c_ft                = p->c_ft;
	pdu->llctrl.cis_req.p_ft                = p->p_ft;
	pdu->llctrl.cis_req.iso_interval        = p->iso_interval;
	pdu->llctrl.cis_req.conn_event_count    = p->conn_event_count;
	(void)memcpy(pdu->llctrl.cis_req.c_max_sdu_packed, p->c_max_sdu_packed,
	       sizeof(p->c_max_sdu_packed));
	(void)memcpy(pdu->llctrl.cis_req.p_max_sdu, p->p_max_sdu, 
		     sizeof(p->p_max_sdu));
	(void)memcpy(pdu->llctrl.cis_req.c_sdu_interval, p->c_sdu_interval, 
		     sizeof(p->c_sdu_interval));
	(void)memcpy(pdu->llctrl.cis_req.p_sdu_interval, p->p_sdu_interval, 
		     sizeof(p->p_sdu_interval));
	(void)memcpy(pdu->llctrl.cis_req.sub_interval, p->sub_interval, 
		     sizeof(p->sub_interval));
	(void)memcpy(pdu->llctrl.cis_req.cis_offset_min, p->cis_offset_min, 
		     sizeof(p->cis_offset_min));
	(void)memcpy(pdu->llctrl.cis_req.cis_offset_max, p->cis_offset_max, 
		     sizeof(p->cis_offset_max));
```

erbr-ot replied:
Prefer to keep line length of max 100 as per guide line
Also, skipping (void) to keep style

Thalley replied:
The guide line is to prefer 80, but will allow up to 100 where needed for better readability

erbr-ot replied:
Well, fair enough re. guide line, however here I much prefer a long line for readability

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -402,6 +407,79 @@ void helper_node_encode_cte_rsp(struct node_rx_pdu *rx, void *param)
+	pdu->len = offsetof(struct pdu_data_llctrl, cis_rsp) +
+		   sizeof(struct pdu_data_llctrl_cis_rsp);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_CIS_RSP;
+	memcpy(pdu->llctrl.cis_rsp.cis_offset_min, p->cis_offset_min, sizeof(p->cis_offset_min));
+	memcpy(pdu->llctrl.cis_rsp.cis_offset_max, p->cis_offset_max, sizeof(p->cis_offset_max));
```
Ditto w.r.t. line length and `(void)`

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -402,6 +407,79 @@ void helper_node_encode_cte_rsp(struct node_rx_pdu *rx, void *param)
 
+void helper_pdu_encode_cis_req(struct pdu_data *pdu, void *param)
+{
+	struct pdu_data_llctrl_cis_req *p = param;
```
```suggestion
	const struct pdu_data_llctrl_cis_req *p = param;
```

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -964,3 +1042,162 @@ void helper_pdu_ntf_verify_cte_rsp(const char *file, uint32_t line, struct pdu_d
+void helper_node_verify_cis_established(const char *file, uint32_t line, struct node_rx_pdu *rx,
+					void *param)
+{
+	struct node_rx_conn_iso_estab *p = (struct node_rx_conn_iso_estab *)param;
+	struct node_rx_conn_iso_estab *pdu = (struct node_rx_conn_iso_estab *)rx->pdu;
```
```suggestion
	const struct node_rx_conn_iso_estab *p = (struct node_rx_conn_iso_estab *)param;
	const struct node_rx_conn_iso_estab *pdu = (struct node_rx_conn_iso_estab *)rx->pdu;
```

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -964,3 +1042,162 @@ void helper_pdu_ntf_verify_cte_rsp(const char *file, uint32_t line, struct pdu_d
+
+void helper_pdu_verify_cis_req(const char *file, uint32_t line, struct pdu_data *pdu, void *param)
+{
+	struct pdu_data_llctrl_cis_req *p = param;
```
```suggestion
	const struct pdu_data_llctrl_cis_req *p = param;
```

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -964,3 +1042,162 @@ void helper_pdu_ntf_verify_cte_rsp(const char *file, uint32_t line, struct pdu_d
+
+void helper_pdu_verify_cis_rsp(const char *file, uint32_t line, struct pdu_data *pdu, void *param)
+{
+	struct pdu_data_llctrl_cis_rsp *p = param;
```
```suggestion
	const struct pdu_data_llctrl_cis_rsp *p = param;
```

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -964,3 +1042,162 @@ void helper_pdu_ntf_verify_cte_rsp(const char *file, uint32_t line, struct pdu_d
+
+void helper_pdu_verify_cis_ind(const char *file, uint32_t line, struct pdu_data *pdu, void *param)
+{
+	struct pdu_data_llctrl_cis_ind *p = param;
```
```suggestion
	const struct pdu_data_llctrl_cis_ind *p = param;
```

Thalley commented on tests/bluetooth/controller/ctrl_cis_create/src/main.c:
```diff
@@ -0,0 +1,468 @@
+/*
+ * Copyright (c) 2020 Demant
+ * Copyright (c) 2020 Nordic Semiconductor ASA
```
```suggestion
 * Copyright (c) 2022 Demant
```

Thalley commented on tests/bluetooth/controller/ctrl_cis_create/src/main.c:
```diff
@@ -0,0 +1,468 @@
+		.cis_id = 0x02
+	};
+	struct pdu_data_llctrl_reject_ext_ind local_reject = {
+		.error_code = 0x17,
```
Is this an HCI error code? If so, please use the macros from `hci_err.h`

erbr-ot replied:
Here it is just a number that should match between the call to reject and the contents of the PDU

Thalley replied:
But shouldn't it be defined somewhere then, so that it is easier to match?

erbr-ot replied:
Will do exactly that

Thalley reviewed (changes requested):
Mostly just formatting and code style comments, otherwise LGTM

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -788,6 +810,33 @@ uint8_t ull_cp_terminate(struct ll_conn *conn, uint8_t error_code)
+{
+	struct proc_ctx *ctx;
+
+	/* TODO
+	 * Check for allowance and parameters
+	 */
```
removing todo

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1034,6 +1083,51 @@ void ull_cp_cte_req_set_disable(struct ll_conn *conn)
+	if (ctx && ctx->proc == PROC_CIS_CREATE) {
+		return ctx->data.cis_create.cis_handle;
+	}
+	return 0xFFFF;
```
Not really - one has been introduced in lll.h, but that should not be used here. Elsewhere 0xffff is used as invalid conn/iso handle

Thalley replied:
We shouldn't use magic numbers (like 0xFFFF) that has special meaning. Suggest to create a new `#define` if you don't think `LLL_HANDLE_INVALID` is fitting here.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -1034,6 +1083,51 @@ void ull_cp_cte_req_set_disable(struct ll_conn *conn)
+
+}
+
+uint16_t ull_cp_cc_ongoing_handle(struct ll_conn *conn)
```
I guess, but I'm following style of existing declarations

Thalley replied:
If you planning on updating existing declaration later, then I'm OK with it.

erbr-ot commented on tests/bluetooth/bsim_bt/edtt_ble_test_app/tests_scripts/gap.llcp.test_list:
```diff
@@ -26,9 +26,9 @@ GAP/CONN/GCEP/BV-05-C
 GAP/CONN/GCEP/BV-06-C
 GAP/CONN/CPUP/BV-01-C
 GAP/CONN/CPUP/BV-02-C
-GAP/CONN/CPUP/BV-03-C
```
Nothing, will move to separate PR

erbr-ot commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -402,6 +407,79 @@ void helper_node_encode_cte_rsp(struct node_rx_pdu *rx, void *param)
+	pdu->len = offsetof(struct pdu_data_llctrl, cis_ind) +
+		   sizeof(struct pdu_data_llctrl_cis_ind);
+	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_CIS_IND;
+	memcpy(pdu->llctrl.cis_ind.aa, p->aa, sizeof(p->aa));
+	memcpy(pdu->llctrl.cis_ind.cis_offset, p->cis_offset, sizeof(p->cis_offset));
+	memcpy(pdu->llctrl.cis_ind.cig_sync_delay, p->cig_sync_delay, sizeof(p->cig_sync_delay));
+	memcpy(pdu->llctrl.cis_ind.cis_sync_delay, p->cis_sync_delay, sizeof(p->cis_sync_delay));
```
Ditto

erbr-ot commented on tests/bluetooth/controller/ctrl_cis_create/src/main.c:
```diff
@@ -0,0 +1,468 @@
+	ut_rx_q_is_empty();
+
+	/* Decline request */
+	ull_cp_cc_reject(&conn, 0x17);
```
see above

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_features.h:
```diff
@@ -132,22 +132,22 @@ static inline bool feature_cte_req(struct ll_conn *conn)
 }
 
 static inline bool feature_smi_tx(struct ll_conn *conn)
 {
-	return LL_FEAT_BIT_SMI_TX;
+	return LL_FEAT_BIT_SMI_TX != 0;
 }
```
yeap

erbr-ot committed:
bluetooth: controller: Fixing feature API re. non standard bool typedef

When bool is typedef'ed to for instance char the cast from 64 bits
is not so good. Now using an explicit 'cast' to true/false

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

Thalley commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -402,6 +407,79 @@ void helper_node_encode_cte_rsp(struct node_rx_pdu *rx, void *param)
+	pdu->llctrl.cis_req.p_ft             =  p->p_ft;
+	pdu->llctrl.cis_req.iso_interval     =  p->iso_interval;
+	pdu->llctrl.cis_req.conn_event_count = p->conn_event_count;
+	memcpy(pdu->llctrl.cis_req.c_max_sdu_packed, p->c_max_sdu_packed,
+	       sizeof(p->c_max_sdu_packed));
+	memcpy(pdu->llctrl.cis_req.p_max_sdu, p->p_max_sdu, sizeof(p->p_max_sdu));
+	memcpy(pdu->llctrl.cis_req.c_sdu_interval, p->c_sdu_interval, sizeof(p->c_sdu_interval));
+	memcpy(pdu->llctrl.cis_req.p_sdu_interval, p->p_sdu_interval, sizeof(p->p_sdu_interval));
+	memcpy(pdu->llctrl.cis_req.sub_interval, p->sub_interval, sizeof(p->sub_interval));
+	memcpy(pdu->llctrl.cis_req.cis_offset_min, p->cis_offset_min, sizeof(p->cis_offset_min));
+	memcpy(pdu->llctrl.cis_req.cis_offset_max, p->cis_offset_max, sizeof(p->cis_offset_max));
```
Ditto with `(void)` and recommended max length of 80.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_peripheral_iso.c:
```diff
@@ -55,25 +66,62 @@ uint8_t ll_cis_accept(uint16_t handle)
+	return acl_conn;
+}
+
+uint8_t ll_cis_accept(uint16_t handle)
+{
+	uint8_t status = BT_HCI_ERR_SUCCESS;
+	struct ll_conn *acl_conn = ll_cis_get_acl_awaiting_reply(handle, &status);
+
+	if (acl_conn) {
+		/* Accept request */
+#if defined(CONFIG_BT_LL_SW_LLCP_LEGACY)
+		acl_conn->llcp_cis.req++;
+#else
+		ull_cp_cc_accept(acl_conn);
+#endif
+	}
+
+	return status;
```
status is updated with proper error code in case acl_conn i returned NULL

Thalley replied:
Ah, yes, sorry missed that :)

Thalley reviewed (dismissed):
Seems OK to me. Some minor changes requested aren't handled, but they are mostly nitpick. 

Personally I would much prefer if the PR followed the official Zephyr coding guidelines w.r.t line length, and the convention of using `(void)` for the `memXXX` functions return value, but those are pretty superficial. The same goes with the use of `const` - It isn't necessary, but I do find it odd to not wanting to do it properly when adding the code, rather than wanting to make another PR later to fix something as trivial as that, later. 

In any case, I'd rather see this go in as is, than not getting in :D

erbr-ot committed:
bluetooth: controller: CIS Terminate Procedure PDU flow and unittest

Implementing PDU flow and unittest for PDU flow of CIS Terminate.

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
bluetooth: controller: Implements Peripheral CIS Create Procedure

Implements PDU flow and unittest of CIS Create on Peripheral
Hooks CIS Create into the controller and fixes a few minor
things in ull to allow for running with NEW LLCP
Also handles CONFIG based comilation of CIS Create

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
bluetooth: controller: Integrate CIS Terminate with controller

Hooks the CIS terminate into the controller and fixes a few minor
things in ull_conn(_iso).c
Also handles CONFIG based comilation of CIS_TERMINATE
also fixes a minor issue in helpers_pdu

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot commented:
fixed CI

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -26,9 +26,11 @@ enum llcp_proc {
 	PROC_CTE_REQ,
-	/* A helper enum entry, to use in pause procedure context */
+	PROC_CIS_TERMINATE,
+	/* A helper enum entry, to use in pause prcedure context */
```
I found it

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_remote.c:
```diff
@@ -776,6 +789,7 @@ static const struct proc_role new_proc_lut[] = {
 	[PDU_DATA_LLCTRL_TYPE_CTE_REQ] = { PROC_CTE_REQ, ACCEPT_ROLE_BOTH },
 #endif /* CONFIG_BT_CTLR_DF_CONN_CTE_RSP */
 	[PDU_DATA_LLCTRL_TYPE_CTE_RSP] = { PROC_UNKNOWN, ACCEPT_ROLE_NONE },
+	[PDU_DATA_LLCTRL_TYPE_CIS_TERMINATE_IND] = { PROC_CIS_TERMINATE, ACCEPT_ROLE_BOTH },
```
Yeap

erbr-ot commented on tests/bluetooth/controller/ctrl_cis_terminate/src/main.c:
```diff
@@ -0,0 +1,159 @@
+		      "Free CTX buffers %d", ctx_buffers_free());
+}
+
+void test_cis_terminate_mas_rem(void)
```
yes

erbr-ot commented on tests/bluetooth/controller/ctrl_cis_terminate/src/main.c:
```diff
@@ -0,0 +1,159 @@
+	cis.cis_id = local_cis_terminate_ind.cis_id;
+	cis.group = &group;
+
+	/* Initiate an LE Ping Procedure */
```
Yes?

erbr-ot replied:
I know ;-)
